utils/parse.js

```diff
@@ -104,18 +104,29 @@ exports.default = function parse(path, content, context) {
     if (!ast || typeof ast !== 'object') {
       console.warn(
         '`parseForESLint` from parser `' +
-          parserPath +
+          parserOrPath + // Can only be invalid for custom parser per imports/parser
           '` is invalid and will just be ignored'
       );
     } else {
-      return makeParseReturn(ast, keysFromParser(parserPath, parser, undefined));
+      return makeParseReturn(ast, keysFromParser(parserOrPath, parser, undefined));
     }
   }
 
   const ast = parser.parse(content, parserOptions);
-  return makeParseReturn(ast, keysFromParser(parserPath, parser, undefined));
+  return makeParseReturn(ast, keysFromParser(parserOrPath, parser, undefined));
 };
 
+function getParser(path, context) {
+  const parserPath = getParserPath(path, context);
+  if (parserPath) {
+      return parserPath;
+  }
+  const isFlat = context.languageOptions &&
+                 typeof context.languageOptions.parser === 'object'; // May be a string
```

Comment:
will this test still work with the approach we've taken to have a config shared between flat and legacy? Meaning, the legacy config object still has `languageOptions`, I just mark it non-enumerable.

Reply:
As far as I understand it is that has nothing to do with the read config file, as `context.languageOptions.parserOptions`  is also set for legacy configs ([relevant code](https://github.com/eslint/eslint/blob/5c2a1986bfc8a01952ed3db617973769fa060494/lib/linter/linter.js#L687)). But if the linter runs in "flat" mode it no longer sets the `context.parserPath`, `context.parserOptions` ... .

Reply:
ok - it's not that languageOptions is _set_ that IDs it as the flat config, it's that it's set and has an object `parser` property?

Are we sure it can't be null, only a string or an object?

Reply:
No as I understand it in the current eslint both properties are set: `context.parserOptions` and `context.languageOptions.parserOptions` (same with other properties of languageOptions). But in "flat" mode the top-level properties are not set.

To the type of `languageOptions.parser`: 
- null case I created an extra test case.
- string: is from [@types/eslint](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a7732c024ce77bcee36cea88228cf4bb210f5da7/types/eslint/index.d.ts#L884)
- object: is already an object for the current eslint

Reply:
Let's also add runtime checks that throw errors if the types aren't what we expect.

Reply:
Added runtime checks that `context.languageOptions.parser` either has a function parse or parseForESLint